File: projects/didi-createorder-msg/src/test/java/com/robinfood/usecases/createorder/CreateOrderUseCaseTest.java
package com.robinfood.usecases.createorder;

import com.robinfood.datamock.OrderToCreateDTOMock;
import com.robinfood.datamock.request.APIGatewayRequestEventMock;
import com.robinfood.dtos.request.ordertocreatedto.OrderToCreateDTO;
import com.robinfood.repository.createtransaction.CreateTransactionRepository;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyMap;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.clearAllCaches;
import static org.mockito.Mockito.when;

class CreateOrderUseCaseTest {


    @Mock
    private CreateTransactionRepository createTransactionRepository;


    @BeforeEach
    public void setup() {
        MockitoAnnotations.openMocks(this);

    }

    @Test
    void test_CreateOrderUseCase_When_DataOk_Should_OK() {

        when(createTransactionRepository.createTransaction(anyMap(), any(), anyString()))
                .thenReturn(OrderToCreateDTOMock.getDataDefault());

        CreateOrderUseCase createOrderUseCase = new CreateOrderUseCase(createTransactionRepository);

        OrderToCreateDTO transactionDTO = createOrderUseCase
                .invoke(
                        OrderToCreateDTOMock.getDataDefault(),
                        "Token",
                        APIGatewayRequestEventMock.getHeaders()
                );

        Assertions.assertEquals(transactionDTO.getPaid(), Boolean.TRUE);
    }

    @Test
    void test_CreateOrderUseCase_Should_BuildConstructor_When_MethodInvoke() {

        CreateOrderUseCase createOrderUseCase = new CreateOrderUseCase();

        Assertions.assertNotNull(createOrderUseCase);
        clearAllCaches();
    }

}

File: projects/posv2/app/src/test/kotlin/com/robinfood/app/usecases/createtransaction/CreateOrderUseCaseTest.kt
package com.robinfood.app.usecases.createtransaction

import com.robinfood.app.mocks.OrderRequestDTOsMocks
import com.robinfood.app.usecases.validateorigin.IValidateOriginUseCase
import com.robinfood.app.usecases.validateuuid.IValidateUuidUseCase
import com.robinfood.core.dtos.OrderCreatedDTO
import com.robinfood.core.dtos.transactionresponse.TransactionCreatedResponseDTO
import com.robinfood.core.dtos.transactionresponse.TransactionResponseDTO
import com.robinfood.core.enums.Result
import com.robinfood.core.mappers.toTransactionRequestEntity
import com.robinfood.repository.transactions.ITransactionRepository
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import java.math.BigDecimal

@ExtendWith(MockitoExtension::class)
class CreateOrderUseCaseTest {

    @Mock
    private lateinit var transactionRepository: ITransactionRepository

    @Mock
    private lateinit var validateOriginUseCase: IValidateOriginUseCase

    @Mock
    private lateinit var validateUuidUseCase : IValidateUuidUseCase

    @InjectMocks
    private lateinit var createTransactionUseCase: CreateTransactionUseCase

    private val token = "token"

    private val orderRequestDTOsMocks = OrderRequestDTOsMocks()

    private val orderRequestDTO = orderRequestDTOsMocks.transactionRequest

    private val orderRequestEntity = orderRequestDTO.toTransactionRequestEntity()

    private val orderCreatedDTOS = listOf(
        OrderCreatedDTO(
            discountPrice = BigDecimal.ZERO,
            id =  1,
            orderInvoiceNumber = "123",
            orderNumber = "123",
            taxPrice = BigDecimal.ZERO,
            total = BigDecimal.valueOf(10000.0),
            subtotal = BigDecimal.ZERO,
            uid = "1234"
        )
    )

    private val orderCreatedResponseDTO = TransactionCreatedResponseDTO(
        transaction = TransactionResponseDTO(
            1L,
            orders = orderCreatedDTOS,
            uuid = "1234"
        )
    )

    @Test
    fun test_CreateOrder_when_service_returns_OK() {
        runBlocking {
            Mockito.`when`(
                transactionRepository.createTransaction(
                    token,
                    orderRequestEntity
                )
            ).thenReturn(Result.Success(orderCreatedResponseDTO))

            Mockito.`when`(
                    validateOriginUseCase.invoke(
                            orderRequestDTO
                    )
            ).thenReturn(orderRequestDTO)

            Mockito.`when`(
                    validateUuidUseCase.invoke(
                            orderRequestDTO
                    )
            ).thenReturn(orderRequestDTO)

            val result = createTransactionUseCase.invoke(
                token,
                orderRequestDTO
            )
            assertEquals(Result.Success(orderCreatedResponseDTO), result)
            verify(validateOriginUseCase).invoke(orderRequestDTO)
        }
    }
}